### PR TITLE
Fix assignment of BMHs for individual compute bmhLabelSelectors

### PR DESCRIPTION
**Pull request by abays**
Our current implementation of assigning computes-to-BMHs based on the formers' `bmhLabelSelector`s does not work properly.  While testing the current form earlier this week, I missed a few cases that would have indicated that there was a problem (and I subsequently ran into them while testing the operator for other reasons today).

Essentially, the same BMH can be matched (and assigned) to multiple computes, because a BMH is not removed from the candidate pool once a compute-to-BMH assignment has been made.  We can't allow this, though, because the mappings must be 1-to-1.  What effectively could happen in such a case was that one or more of the `selectedBaremetalHosts` mappings was overwritten by whichever computes happened to be processed last.  Thus the `selectedBaremetalHosts` map would have fewer entries than the needed number of mappings, which resulted in the create/update of the `OpenStackBaremetalSet` being rejected in the webhook -- even if there was a particular set of mappings of computes-to-BMHs that would have worked for the labels in question.

The algorithm for mapping computes-to-BMHs has been rewritten to consider all possible matches until it finds a set that satisfies all the computes involved.  If it exhausts all possible combinations, it returns a failure.

### CODE REVIEW

**Comment by abays** [comment]
recheck

```
RETRY_LIMIT
```

**Comment by abays** [comment]
> @abays: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> Test name 	Commit 	Details 	Required 	Rerun command
> ci/prow/openstack-baremetal-operator-build-deploy 	[3a2ce23](https://github.com/openstack-k8s-operators/openstack-baremetal-operator/commit/3a2ce231672331f8614ec28d2f4abec03f4d3282) 	[link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openstack-k8s-operators_openstack-baremetal-operator/218/pull-ci-openstack-k8s-operators-openstack-baremetal-operator-main-openstack-baremetal-operator-build-deploy/1841912114273325056) 	false 	`/test openstack-baremetal-operator-build-deploy`
> 
> [Full PR test history](https://prow.ci.openshift.org/pr-history?org=openstack-k8s-operators&repo=openstack-baremetal-operator&pr=218). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:abays).

```
{  error occurred handling build src-amd64: could not get build src-amd64: builds.build.openshift.io "src-amd64" not found}
```

/test openstack-baremetal-operator-build-deploy

File: api/v1beta1/openstackbaremetalset.go [hunk 1]
```diff
@@ -144,12 +151,60 @@ func VerifyBaremetalSetScaleUp(
...
+		// We only want to continue to individual compute label matching if we actually have
+		// enough BMHs remaining given the filtering above
+		if len(availableBaremetalHosts) >= newBmhsNeededCount {
+
+			// Now try to fit all the requested new compute hosts into the set of available BMHs
+			// given bmhLabelSelectors on each compute (if any) and the labels on the BMHs
+			//
+			// The problem we are trying to solve can be demonstrated through an example...
+			// Imagine we have 3 available BMHs at this point with the following simplified labels:
+			//
```
Comment (rabi) [inline api/v1beta1/openstackbaremetalset.go:168]: I think the assumption I had with the existing logic is that labels for individual node selection would be unique and not overlapping  (i.e not have more than one BMH matching the same labels) and we will always find the first label match like k8s.

File: api/v1beta1/openstackbaremetalset.go [hunk 2]
```diff
@@ -190,19 +246,81 @@ func VerifyBaremetalSetScaleDown(
...
+	computesToBmhs := map[string][]metal3v1.BareMetalHost{}
+	for compName, comp := range computes {
+		for _, bmh := range bmhs {
+			if IsMapSubset(bmh.GetLabels(), comp.BmhLabelSelector) {
+				computesToBmhs[compName] = append(computesToBmhs[compName], bmh)
+			}
+		}
+	}
+
+	// Now sort computes by the number of valid matches, as sorting helps in
```
Comment (rabi) [inline api/v1beta1/openstackbaremetalset.go:261]: Do we need to all this? Can we not just sort the hosts in the `BaremetalHosts` and BMHs based on their labels and then we would always find the correct match?

Comment (rabi) [reply api/v1beta1/openstackbaremetalset.go:261]: Do you think sorting like https://github.com/openstack-k8s-operators/openstack-baremetal-operator/compare/main...rabi:openstack-baremetal-operator:bmh_selection?expand=1  would work for the scenarios you're tesing? Should we add some functional testing for those scenarios?

Comment (abays) [reply api/v1beta1/openstackbaremetalset.go:261]: Yes.  I built your fork/branch and used it with the same series of manual tests I was trying, and everything worked as expected.  I'm not partial to my PR and would be just as comfortable with using your approach instead.  Basically I'm just interest it getting this working properly.

Here is an example of what happens currently.  I have an `OpenStackBaremetalSet` without any `bmhLabelSelector`s:

```
...
  baremetalHosts:
    edpm-compute-0: 
      ctlPlaneIP: 192.168.122.100
    edpm-compute-1: 
      ctlPlaneIP: 192.168.122.101
    edpm-compute-2:
      ctlPlaneIP: 192.168.122.102
  #bmhLabelSelector:
  #  app: openstack
...
```

When I go to create it, I see this:

```
$ oc apply -f baremetalset.yaml 
Error from server (Forbidden): error when creating "baremetalset.yaml": admission webhook "vopenstackbaremetalset.kb.io" denied the request: unable to find 3 requested BaremetalHosts in namespace openshift-machine-api for scale-up (0 in use, 2 available)

$ oc apply -f baremetalset.yaml 
Error from server (Forbidden): error when creating "baremetalset.yaml": admission webhook "vopenstackbaremetalset.kb.io" denied the request: unable to find 3 requested BaremetalHosts in namespace openshift-machine-api for scale-up (0 in use, 1 available)
```

You can see in the output `(0 in use, X available)` messages that the randomness of computes overwriting each other for BMH assignment is happening.  Anyhow, it seems like we're converging on a solution.

**Review by rabi — APPROVED** [review]
/lgtm

**Pull request merged**